app/src/backend/hiap/HiapService.ts

```diff
@@ -1039,20 +1091,69 @@ async function copyRankedActionsToLang(ranking: any, lang: LANGUAGES) {
     (a, b) => a.rank - b.rank,
   );
 
+  // Aggregate available languages across all actions
+  const availableLanguagesSet = new Set<string>();
+  for (const action of uniqueActions) {
+    if (action.explanation && typeof action.explanation === 'object') {
+      Object.keys(action.explanation).forEach((lang) =>
+        availableLanguagesSet.add(lang),
+      );
+    }
+  }
+  const availableLanguages = Array.from(availableLanguagesSet).sort();
+  const hasRequestedLang = availableLanguages.includes(lang);
+
   logger.info(
+    {
+      requestedLang: lang,
+      availableLanguages,
+      rankingLangs: ranking.langs,
+      hasRequestedLang,
+      actionCount: uniqueActions.length,
+    },
     `Copying ${uniqueActions.length} unique actions to language ${lang}`,
   );
 
+  if (!hasRequestedLang) {
+    logger.warn(
+      {
+        requestedLang: lang,
+        availableLanguages,
+        rankingId: ranking.id,
+      },
+      `⚠️  Requested language ${lang} not found in existing explanations. ` +
+        `Explanations will only contain: ${availableLanguages.join(", ")}. ` +
+        `Frontend should fall back to an available language.`,
+    );
+  }
+
   const rankedActions = uniqueActions.map((r) => ({
     actionId: r.actionId,
     rank: r.rank,
-    explanation: r.explanation,
+    explanation: r.explanation, // Pass through explanation object with available languages
     type: r.type as ACTION_TYPES,
   }));
 
-  // Fetch and merge action details in the requested language
+  // Fetch and merge action details in the requested language (German names/descriptions)
```

Comment:
This comment seems off, I think it's the selected language not always German right?